File: algorithms/src/main/java/com/arpan/dsa/datastructures/linkedlist/LinkedList.java
package com.arpan.dsa.datastructures.linkedlist;

public class LinkedList {

    private LinkedListNode head;

    private LinkedListNode tail;

    public LinkedListNode search(int key) {
        LinkedListNode curLinkedListNode = head;

        while (curLinkedListNode != null) {
            if (curLinkedListNode.data == key) {
                return curLinkedListNode;
            }
            curLinkedListNode = curLinkedListNode.next;
        }
        return null;
    }

    public void insert(int data) {
        LinkedListNode linkedListNode = new LinkedListNode(data);
        if (head == null) {
            head = linkedListNode;
            tail = linkedListNode;
        } else {
            tail.next = linkedListNode;
            linkedListNode.prev = tail;
            tail = linkedListNode;
        }
    }

    public void delete(LinkedListNode linkedListNode) {
        if (linkedListNode != null) {
            if (linkedListNode.prev != null) {
                linkedListNode.prev.next = linkedListNode.next;
            } else {
                head = linkedListNode.next;
            }
            if (linkedListNode.next != null) {
                linkedListNode.next.prev = linkedListNode.prev;
            } else {
                tail = linkedListNode.prev;
            }
        }
    }

    @Override
    public String toString() {
        LinkedListNode node = head;
        StringBuilder str = new StringBuilder();
        str.append("LinkedList: [");
        while (node != null) {
            str.append(node.data).append(" ->");
            node = node.next;
        }
        str.append(" ]");
        return str.toString();
    }
}

File: grokking-coding-interview-patterns/05_linked_list_manipulation/linked_list.py
from typing import Optional

from linked_list_node import LinkedListNode


# Template for the linked list
class LinkedList:

    # __init__ will be used to make a LinkedList type object.
    def __init__(self):
        self.head: Optional[LinkedListNode] = None

    # insert_node_at_head method will insert a LinkedListNode at
    # head of a linked list.
    def insert_node_at_head(self, node: LinkedListNode):
        if self.head:
            node.next = self.head
            self.head = node
        else:
            self.head = node

    # create_linked_list method will create the linked list using the
    # given integer array with the help of InsertAthead method.
    def create_linked_list(self, lst: list):
        for x in reversed(lst):
            new_node = LinkedListNode(x)
            self.insert_node_at_head(new_node)

    # __str__(self) method will display the elements of linked list.
    def __str__(self):
        result = ""
        temp = self.head
        while temp:
            result += str(temp.data)
            temp = temp.next
            if temp:
                result += ", "
        result += ""
        return result


def print_list(head: Optional[LinkedListNode], prefix: str):
    curr = head

    print(f"{prefix}: ", end="")
    while curr is not None:
        if curr.next is None:
            print(f"{curr.data}", end="")
        else:
            print(f"{curr.data} -> ", end="")
        curr = curr.next
    print()
